### PR TITLE
Authenticate internal connections via domain sockets

### PR DESCRIPTION
**Pull request by adam8157**
In `cdbconn_doConnectStart()`, Greenplum cleans the host info to force
using domain socket for entry db connections. Checking the connection's
authentication here is not necessary, even annoying, if the user is not
allowed to connect via domain socket.

This commit authenticates it directly, just as Greenplum authenticates
TCP internal connections coming from the same host.

Fix https://github.com/greenplum-db/gpdb/issues/12851

### CODE REVIEW

**Comment by adam8157** [comment]
From the security perspective, I think authenticating internal connections is just like authenticating TCP internal connections coming from the same host. Is it?

File: src/backend/libpq/auth.c [hunk 1]
```diff
@@ -529,29 +529,35 @@ internal_client_authentication(Port *port)
...
-			   )
+		   )
 		{
-			if (check_same_host_or_net(&port->raddr, ipCmpSameHost) &&
-				!gp_reject_internal_tcp_conn)
+			if (check_same_host_or_net(&port->raddr, ipCmpSameHost))
 			{
-				elog(DEBUG1, "received same host internal TCP connection");
-				FakeClientAuthentication(port);
+				if (!gp_reject_internal_tcp_conn)
```
Comment (ashwinstar) [inline src/backend/libpq/auth.c:550]: Please let's avoid the negation in the condition, much simpler to read code without it

File: src/backend/libpq/auth.c [hunk 2]
```diff
@@ -529,29 +529,35 @@ internal_client_authentication(Port *port)
...
+				{
+					elog(DEBUG1, "received same host internal TCP connection");
+					FakeClientAuthentication(port);
```
Comment (ashwinstar) [inline src/backend/libpq/auth.c:553]: best to move the `return true` after this and remove it from top-level `if` block

File: src/backend/libpq/auth.c [hunk 3]
```diff
@@ -529,29 +529,35 @@ internal_client_authentication(Port *port)
...
+				}
+				else
+				{
+					elog(DEBUG1, "rejecting TCP connection to master using internal"
+						 "connection protocol, because the GUC gp_reject_internal_tcp_conn is true");
+					return false;
+				}
 			}
 			else
```
Comment (ashwinstar) [inline src/backend/libpq/auth.c:562]: This `else` won't be needed once we move the `return` inside

File: src/backend/libpq/auth.c [hunk 4]
```diff
@@ -529,29 +529,35 @@ internal_client_authentication(Port *port)
...
+			if (check_same_host_or_net(&port->raddr, ipCmpSameHost))
 			{
-				elog(DEBUG1, "received same host internal TCP connection");
-				FakeClientAuthentication(port);
+				if (!gp_reject_internal_tcp_conn)
+				{
+					elog(DEBUG1, "received same host internal TCP connection");
+					FakeClientAuthentication(port);
+				}
+				else
```
Comment (ashwinstar) [inline src/backend/libpq/auth.c:555]: if `return` is added this `else` won't be needed

File: src/backend/libpq/auth.c [hunk 5]
```diff
@@ -565,28 +571,11 @@ internal_client_authentication(Port *port)
 #ifdef HAVE_UNIX_SOCKETS
 		else if (port->raddr.addr.ss_family == AF_UNIX)
 		{
-			/* 
-			 * Internal connection via a domain socket -- use ident
+			/*
+			 * Internal connection via a domain socket -- authenticate it
```
Comment (ashwinstar) [inline src/backend/libpq/auth.c:575]: should we say "consider it authenticated" as we are faking the authentication?

**Review by ashwinstar — APPROVED** [review]
We need to in long run, really improve this situation and avoid all this `FakeClientAuthentication()` business for internal connections. For now, this change seems fine.

**Pull request merged**